packages/nextjs/pages/api/etl/index.ts

```diff
@@ -4,55 +4,61 @@ import type { NextApiRequest, NextApiResponse } from "next";
+import ProjectMetricSummary, { TempProjectMetricSummary } from "~~/services/mongodb/models/projectMetricSummary";
+import ProjectScore, { IProjectScore, TempProjectScore } from "~~/services/mongodb/models/projectScore";
+
+export const maxDuration = 300;
+export const runtime = "nodejs";
 
 export default async function handler(req: NextApiRequest, res: NextApiResponse) {
   if (req.method !== "GET") {
     return res.status(405).json({ message: "Method not allowed." });
   }
-  const weightings = weightingsJSON as { [key in keyof Metrics]: number };
   try {
-    await dbConnect();
+    const mongooseConnection = await dbConnect();
     // Log the ETL process
-    const existingLog = await ETLLog.findOne({ status: "pending" });
+    const existingLog = await ETLLog.findOne({}, {}, { sort: { _id: -1 } });
     if (existingLog) {
-      return res.status(400).json({ message: `ETL Process already running as of ${existingLog.date}` });
+      if (existingLog.status === "pending") {
+        return res.status(400).json({ message: `ETL Process already running as of ${existingLog.date}` });
+      }
+      if (existingLog.status === "success" && isTooEarly(existingLog.date)) {
+        return res.status(400).json({ message: `Too soon to run ETL Process. Last run: ${existingLog.date}` });
+      }
     }
+
     const log = await ETLLog.create({ date: new Date(), status: "pending" });
     res.status(200).json({ result: "ETL Process Started" });
-    // Start a session
-    const session = await startSession();
 
-    // Start a transaction
-    // session.startTransaction();
+    // Get weights from JSON file
+    const weightings = weightingsJSON as { [key in keyof Metrics]: number };
+
     try {
-      // Clear out the previous data
-      console.log("Clearing previous data");
-      await GlobalScore.deleteMany({}, { session });
-      await ProjectScore.deleteMany({}, { session });
-      await ProjectMetricSummary.deleteMany({}, { session });
       // Update new data
       // Get all the mapping data
-      const { mapping } = await fetch("http://localhost:3000/api/stub/mapping").then(res => res.json());
+      const { mapping } = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/stub/mapping`).then(res => res.json());
       // Get metrics that are activated
       const metrics = await Metric.findAllActivated();
       if (!metrics) {
-        return res.status(404).json({ message: "No metrics found" });
+        throw new Error("No metrics found");
       }
+      const metricNamesObj: Partial<Metrics> = metrics.reduce((acc: Partial<Metrics>, metric) => {
+        acc[metric.name] = 0;
+        return acc;
+      }, {} as Partial<Metrics>);
       const dates = getDatesToProcess();
-      const globalScoreData = Object.assign({}, MetricNames) as { [key in keyof Metrics]: number };
+      const globalScoreData = Object.assign({}, metricNamesObj) as { [key in keyof Metrics]: number };
       for (const day of dates) {
```

Comment:
I think parallelizing here would significantly speed things up. I've been messing around with the following

```js
      await Promise.all(
        dates.map(async day => {
          // Fetch data for the day
          const seriesResponse = await fetch(
            `${process.env.NEXT_PUBLIC_API_URL}/stub/series?date=${day.toISOString().split("T")[0]}`,
          ).then(res => res.json());

          if (!seriesResponse) {
            throw new Error(`No OSO Data for ${day}`);
          }
          const osoData = seriesResponse.data as OnchainMetricsByProject[];
          console.log(`Processing OSO response for each project on ${day.toISOString().split("T")[0]}`);

          const projectScoreOps = [];
          const globalScoreOps = [];

          for (const project of osoData) {
            const projectMapping = mapping.find((map: any) => map.oso_name === project.project_name);
            if (!projectMapping) {
              console.error(`No mapping found for ${project.project_name}`);
              continue;
            }
            const projectId = projectMapping.application_id;
            const impact_index = getImpactIndex(project as unknown as Metrics, weightings);

            const projectMetrics = {} as { [key in keyof Metrics]: number };
            for (const metric of metrics) {
              const metricValue = project[metric.name as keyof OnchainMetricsByProject];
              if (!isNaN(metricValue as number)) {
                projectMetrics[metric.name as keyof Metrics] = parseInt(metricValue as string);
              }
            }
            const projectScoreData = Object.assign(projectMetrics, {
              date: day,
              projectId,
              impact_index,
            }) as IProjectScore;
            projectScoreOps.push({
              insertOne: {
                document: projectScoreData,
              },
            });

            for (const metric of Object.keys(metricNamesObj) as (keyof Metrics)[]) {
              if (projectMetrics[metric]) {
                globalScoreData[metric] += projectMetrics[metric];
              }
            }
          }

          globalScoreData.impact_index = getImpactIndex(globalScoreData, weightings);
          globalScoreOps.push({
            insertOne: {
              document: { date: day, ...globalScoreData },
            },
          });

          // Batch insert project scores
          if (projectScoreOps.length > 0) {
            await TempProjectScore.bulkWrite(projectScoreOps);
          }
          // Batch insert global scores
          if (globalScoreOps.length > 0) {
            await TempGlobalScore.bulkWrite(globalScoreOps);
          }
        }),
      );
```

Reply:
This is great!